File: packages/turbo-cache/android/src/main/java/com/reactnativeiconify/turbocache/TurboCacheModule.kt
// Copyright 2025-present React Native Iconify. All rights reserved.

package com.reactnativeiconify.turbocache

import android.content.Context
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.cache.DiskCache
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File

/**
 * TurboCacheModule provides native disk caching using Glide's DiskCache.
 *
 * This module uses Glide's disk cache for persistent caching with automatic LRU eviction.
 * Cache keys are prefixed with "turbo-cache:" to avoid collision with image cache.
 *
 * Based on expo-image's Glide implementation.
 */
class TurboCacheModule : Module() {
  companion object {
    private const val CACHE_KEY_PREFIX = "turbo-cache:"
  }

  private val context: Context
    get() = appContext.reactContext ?: throw IllegalStateException("React context is not available")

  private val diskCache: DiskCache
    get() = Glide.get(context).diskCache

  override fun definition() = ModuleDefinition {
    Name("TurboCache")

    // Get value from cache
    AsyncFunction("get") { key: String ->
      getCacheValue(key)
    }

    // Set value to cache with optional TTL
    AsyncFunction("set") { key: String, value: String, ttl: Int? ->
      setCacheValue(key, value, ttl)
    }

    // Remove value from cache
    AsyncFunction("remove") { key: String ->
      removeCacheValue(key)
    }

    // Clear all cache
    AsyncFunction("clear") {
      clearAllCache()
    }

    // Get cache size in bytes
    AsyncFunction("getSize") {
      getCacheSize()
    }

    // Clear memory cache (matching expo-image pattern)
    AsyncFunction("clearMemoryCache") {
      clearMemoryCache()
    }

    // Clear disk cache (matching expo-image pattern)
    AsyncFunction("clearDiskCache") {
      clearDiskCache()
    }
  }

  /**
   * Creates a cache key with prefix to avoid collision with image cache.
   */
  private fun createCacheKey(key: String): DiskCache.Key {
    return object : DiskCache.Key {
      override fun updateDiskCacheKey(messageDigest: java.security.MessageDigest) {
        messageDigest.update((CACHE_KEY_PREFIX + key).toByteArray())
      }
    }
  }

  /**
   * Get value from cache.
   * Returns null if key doesn't exist or data is corrupted/expired.
   */
  private suspend fun getCacheValue(key: String): String? = withContext(Dispatchers.IO) {
    try {
      val cacheKey = createCacheKey(key)
      val file = diskCache.get(cacheKey) ?: return@withContext null

      val jsonString = file.readText()
      val cacheData = Json.decodeFromString<CacheData>(jsonString)

      // Check if expired
      if (cacheData.isExpired()) {
        diskCache.delete(cacheKey)
        return@withContext null
      }

      cacheData.value
    } catch (e: Exception) {
      throw CacheException("Failed to get from cache: ${e.message}", e)
    }
  }

  /**
   * Set value to cache with optional TTL (in milliseconds).
   */
  private suspend fun setCacheValue(key: String, value: String, ttl: Int?) = withContext(Dispatchers.IO) {
    try {
      val cacheKey = createCacheKey(key)

      // Wrap value with expiration timestamp if TTL is provided
      val cacheData = if (ttl != null) {
        val expiresAt = System.currentTimeMillis() + ttl
        CacheData(value, expiresAt)
      } else {
        CacheData(value, null)
      }

      // Encode to JSON
      val jsonString = Json.encodeToString(cacheData)
      val bytes = jsonString.toByteArray()

      // Write to disk cache
      diskCache.put(cacheKey, object : DiskCache.Writer {
        override fun write(file: File): Boolean {
          return try {
            file.writeBytes(bytes)
            true
          } catch (e: Exception) {
            false
          }
        }
      })
    } catch (e: Exception) {
      throw CacheException("Failed to set to cache: ${e.message}", e)
    }
  }

  /**
   * Remove value from cache.
   */
  private suspend fun removeCacheValue(key: String) = withContext(Dispatchers.IO) {
    try {
      val cacheKey = createCacheKey(key)
      diskCache.delete(cacheKey)
    } catch (e: Exception) {
      throw CacheException("Failed to remove from cache: ${e.message}", e)
    }
  }

  /**
   * Clear all cache entries.
   * Note: This clears ALL disk cache, not just turbo-cache entries.
   */
  private suspend fun clearAllCache() = withContext(Dispatchers.IO) {
    try {
      diskCache.clear()
    } catch (e: Exception) {
      throw CacheException("Failed to clear cache: ${e.message}", e)
    }
  }

  /**
   * Get total cache size in bytes.
   */
  private suspend fun getCacheSize(): Long = withContext(Dispatchers.IO) {
    try {
      diskCache.currentSizeBytes
    } catch (e: Exception) {
      throw CacheException("Failed to get cache size: ${e.message}", e)
    }
  }

  /**
   * Clear memory cache (matching expo-image pattern).
   */
  private fun clearMemoryCache(): Boolean {
    try {
      Glide.get(context).clearMemory()
      return true
    } catch (e: Exception) {
      throw CacheException("Failed to clear memory cache: ${e.message}", e)
    }
  }

  /**
   * Clear disk cache (matching expo-image pattern).
   */
  private suspend fun clearDiskCache(): Boolean = withContext(Dispatchers.IO) {
    try {
      Glide.get(context).clearDiskCache()
      true
    } catch (e: Exception) {
      throw CacheException("Failed to clear disk cache: ${e.message}", e)
    }
  }
}

/**
 * Wrapper for cached data with expiration support.
 */
@Serializable
private data class CacheData(
  val value: String,
  val expiresAt: Long? = null
) {
  fun isExpired(): Boolean {
    val expiresAt = expiresAt ?: return false
    return System.currentTimeMillis() > expiresAt
  }
}

/**
 * Custom exception for cache operations.
 */
private class CacheException(message: String, cause: Throwable? = null) : Exception(message, cause)

File: packages/turbo-cache/ios/TurboCacheModule.swift
// Copyright 2025-present React Native Iconify. All rights reserved.

import SDWebImage
import ExpoModulesCore

/**
 * TurboCacheModule provides native disk caching using SDWebImage's SDImageCache.
 *
 * This module uses SDImageCache.shared for persistent caching with automatic LRU eviction.
 * Cache keys are prefixed with "turbo-cache:" to avoid collision with image cache.
 *
 * Based on expo-image's ImageLoader implementation.
 */
public class TurboCacheModule: Module {
  // MARK: - Constants

  private static let cacheKeyPrefix = "turbo-cache:"
  private let cache = SDImageCache.shared

  // MARK: - Module Definition

  public func definition() -> ModuleDefinition {
    Name("TurboCache")

    // Get value from cache
    AsyncFunction("get") { (key: String) -> String? in
      return try await self.getCacheValue(for: key)
    }

    // Set value to cache with optional TTL
    AsyncFunction("set") { (key: String, value: String, ttl: Int?) in
      try await self.setCacheValue(value, for: key, ttl: ttl)
    }

    // Remove value from cache
    AsyncFunction("remove") { (key: String) in
      try await self.removeCacheValue(for: key)
    }

    // Clear all cache
    AsyncFunction("clear") {
      try await self.clearAllCache()
    }

    // Get cache size in bytes
    AsyncFunction("getSize") { () -> Int in
      return try await self.getCacheSize()
    }

    // Clear memory cache (matching expo-image pattern)
    AsyncFunction("clearMemoryCache") { () -> Bool in
      self.cache.clearMemory()
      return true
    }

    // Clear disk cache (matching expo-image pattern)
    AsyncFunction("clearDiskCache") { () -> Bool in
      return try await self.clearDiskCache()
    }
  }

  // MARK: - Private Methods

  /**
   * Creates a cache key with prefix to avoid collision with image cache.
   */
  private func createCacheKey(_ key: String) -> String {
    return Self.cacheKeyPrefix + key
  }

  /**
   * Get value from cache.
   * Returns nil if key doesn't exist or data is corrupted.
   */
  private func getCacheValue(for key: String) async throws -> String? {
    let cacheKey = createCacheKey(key)

    return try await withCheckedThrowingContinuation { continuation in
      // Query cache for data (disk only, memory cache not needed for string data)
      cache.queryCacheOperation(
        forKey: cacheKey,
        context: nil,
        cacheType: .disk
      ) { (_, data, cacheType) in
        guard let data = data else {
          continuation.resume(returning: nil)
          return
        }

        // Convert data to string
        if let value = String(data: data, encoding: .utf8) {
          continuation.resume(returning: value)
        } else {
          continuation.resume(
            throwing: CacheDecodingError(message: "Failed to decode cached data for key: \(key)")
          )
        }
      }
    }
  }

  /**
   * Set value to cache with optional TTL.
   * Note: SDWebImage doesn't support per-key TTL, only global maxDiskAge.
   * For TTL support, we store expiration metadata in the value itself.
   */
  private func setCacheValue(_ value: String, for key: String, ttl: Int?) async throws {
    let cacheKey = createCacheKey(key)

    // Wrap value with expiration timestamp if TTL is provided
    let cacheData: CacheData
    if let ttl = ttl {
      let expiresAt = Date().timeIntervalSince1970 + Double(ttl) / 1000.0
      cacheData = CacheData(value: value, expiresAt: expiresAt)
    } else {
      cacheData = CacheData(value: value, expiresAt: nil)
    }

    // Encode to JSON
    guard let jsonData = try? JSONEncoder().encode(cacheData),
          let jsonString = String(data: jsonData, encoding: .utf8),
          let data = jsonString.data(using: .utf8) else {
      throw CacheEncodingError(message: "Failed to encode value for key: \(key)")
    }

    return try await withCheckedThrowingContinuation { continuation in
      // Store to disk cache only
      cache.store(
        nil,
        imageData: data,
        forKey: cacheKey,
        cacheType: .disk
      ) {
        continuation.resume()
      }
    }
  }

  /**
   * Remove value from cache.
   */
  private func removeCacheValue(for key: String) async throws {
    let cacheKey = createCacheKey(key)

    return try await withCheckedThrowingContinuation { continuation in
      cache.removeImage(
        forKey: cacheKey,
        cacheType: .disk
      ) {
        continuation.resume()
      }
    }
  }

  /**
   * Clear all cache entries with our prefix.
   * Note: This clears ALL disk cache, not just turbo-cache entries.
   * For selective clearing, we'd need to track keys separately.
   */
  private func clearAllCache() async throws {
    return try await withCheckedThrowingContinuation { continuation in
      // Clear disk cache only (memory cache not used)
      cache.clearDisk {
        continuation.resume()
      }
    }
  }

  /**
   * Get total cache size in bytes.
   */
  private func getCacheSize() async throws -> Int {
    return try await withCheckedThrowingContinuation { continuation in
      cache.calculateSize { (fileCount, totalSize) in
        continuation.resume(returning: Int(totalSize))
      }
    }
  }

  /**
   * Clear disk cache (matching expo-image pattern).
   */
  private func clearDiskCache() async throws -> Bool {
    return try await withCheckedThrowingContinuation { continuation in
      cache.clearDisk {
        continuation.resume(returning: true)
      }
    }
  }
}

// MARK: - Helper Types

/**
 * Wrapper for cached data with expiration support.
 */
private struct CacheData: Codable {
  let value: String
  let expiresAt: Double?

  var isExpired: Bool {
    guard let expiresAt = expiresAt else {
      return false
    }
    return Date().timeIntervalSince1970 > expiresAt
  }
}

// MARK: - Errors

private class CacheEncodingError: Exception {
  let message: String

  init(message: String) {
    self.message = message
    super.init()
  }

  override var reason: String {
    return message
  }
}

private class CacheDecodingError: Exception {
  let message: String

  init(message: String) {
    self.message = message
    super.init()
  }

  override var reason: String {
    return message
  }
}
